슬라이드 방향키 이동 구현 완료

aereeeee opened the pull request:
# 슬라이드 방향키 이동 구현 완료

## 해당 이슈 📎

- 슬라이드를 좌우 화살표 키로 이동할 수 있음 (#133 )

## 변경 사항 🛠

- 방향키로 이동 구현
- 채팅 인풋 focus 시 방향키는 커서 이동, 채팅 인풋 창에서 focus out되면 동작
- 렌더링 최적화의 첫걸음으로 현재 채널 context API에 reducer적용

## 테스트 ✨
> 없음

## 리뷰어 참고 사항 🙋‍♀️

> 없음

aereeeee committed:
[ FEATURE ] 좌우 방향키 슬라이드 이동 구현 (#133)

Why
- 방향키로 슬라이드 이동이 가능해야한다

What
- keydown이벤트 발생 시  페이지 이동하는 로직을 구현하였다

aereeeee committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-02 into feature/issue133

aereeeee committed:
[ REFACTOR ] app routing 코드 리팩토링

Why
- routing 코드의 가독성이 좋지 않다.
- routing만 담당하는 컴포넌트로 분리하는 것이 좋다.

What
- route의 component prop사용
- wrapper와 header 즉 레이아웃을 위한 컴포넌트는 base 역할을 하는 Layout컴포넌트로 분리하여 작성

aereeeee committed:
[ FEATURE ] 채널 관련 상태 관리를 위해 context API, reducer활용 provider컴포넌트 구현 (#133)

Why
- 채팅 중인 여부와 같이 채널 전체적으로 관리해야 할 상태가 존재한다.

What
- 기존 그래프큐엘 쿼리 결과를 내려주던 프로바이더를 컴포넌트로 분리하여 리듀서의 데이터도 함께 내려주도록 수정
- dispatch 프로바이더를 별도로 구현하여 렌더링 최적화

aereeeee committed:
[ FEATURE ] 채팅 입력 시 방향키 사용이 슬라이드 이동을 유발하지 않도록 개선 (#133)

Why
- 채팅 입력시에 방향키를 사용하면 커서가 이동해야하고 슬라이드는 이동하지 않아야한다.

What
- 전역으로 사용하는 리듀서의 상태를 이용하여 채팅 입력중이면 방향키로 슬라이드 이동을 막는다

aereeeee committed:
[ REFACTOR ] 구현된 페이지 제외 not found 페이지에 대한 에러메세지 작성

Why
- 존재하지 않는 페이지일 경우 메시지가 표시되어야 한다.

What
- 메시지를 상수로 분리하여 에러 모달에 전달하도록 수정

happydhKim commented on web/src/components/channel/Chat/ChatInput/ChatInput.jsx:
```diff
@@ -25,13 +26,21 @@ const ChatInput = (props) => {
   };
+  const handleFocus = (isChat) => () => {
+    dispatch({
+      type: 'SET_ISCHAT',
+      payload: { isChat },
```
두개 니까 한줄로 하는게 어떨까요? 우리는 컨벤션이 세개잖아욥?🌝

happydhKim commented on web/src/components/channel/Slide/SlideViewer/SlideViewer.jsx:
```diff
@@ -13,6 +13,9 @@ import Indicator from './Indicator';
 import PageNumber from './PageNumber';
 
+const KEYCODE_BACK = 37;
+const KEYCODE_FOWARD = 39;
```
매직넘버로 보여요 constants로 빼볼까요??😀

aereeeee replied:
keycode는 전체적으로 상단에 상수로 빼주는걸로 합의된것 아닌가요??
아니엇군요ㅎㅎ constants로 빼겠습니다.

happydhKim commented on web/src/components/channel/Slide/SlideViewer/SlideViewer.jsx:
```diff
@@ -25,13 +28,17 @@ const SlideViewer = (props) => {
   );
+  const directionKey = {};
+  directionKey[KEYCODE_BACK] = 'back';
+  directionKey[KEYCODE_FOWARD] = 'foward';
```
```javascript
const KEYCODE_BACK = 37;
const KEYCODE_FOWARD = 39;
```

이것까지 같이 정리하면 좋을거 같아요☺️

happydhKim commented on web/src/components/channel/Slide/SlideViewer/SlideViewer.jsx:
```diff
@@ -42,14 +49,26 @@ const SlideViewer = (props) => {
+  const handleKeyDown = useCallback((event) => {
+    if (isChat) return;
+    const key = event.keyCode;
+    if (Object.keys(directionKey).includes(key.toString())) {
```
오브젝트 키값을 빼는 과정과 그것을 포함하는 함수 그리고 toString까지 있으니
함수로 빼는게 좋지 않을까요?🤭

aereeeee replied:
키다운 핸들러에서만 쓰는 함수를 컴포넌트 내에 만들고 싶지 않아서 고민되네요.. 일단 머지하고 추후에 생각해보겠습니다!

happydhKim commented on web/src/components/channel/Slide/SlideViewer/SlideViewer.jsx:
```diff
@@ -42,14 +49,26 @@ const SlideViewer = (props) => {
 
   const screenChange = (e) => setFullScreen(e);
-  const IndicatorRender = ['back', 'foward'].map((direction) => (
+  const IndicatorRender = Object.values(directionKey).map((direction) => (
```
이부분도 map을 하는 과정과 jsx문법을 나눠서 처리하면 어떨까요?☺️

aereeeee replied:
컴포넌트 생성 로직을 간단히 하기 위해 map함수를 사용하고있고, Object.values(directionKey) 부분만 변수로 빼는 것은 좀 애매한것 같은데요, 이미 directionKey도  그런 이유로 변수로 빼주었고.. 일단 두겠습니다!

happydhKim reviewed (approved):
드디어 기다리던 기능이 완성되었어요!🤭
너무너무 감사하고 고생하셨어요!😎

aereeeee committed:
[ REFACTOR ] 리뷰 사항 반영 (#133)

Why
- 클린한 코드를 위해 상수를 빼는 편이 좋다.

What
- 방향키 코드를 상수로 뺌
- back, foward 상태를 true false로 변경
